src/backend/distributed/master/master_delete_protocol.c

```diff
@@ -359,99 +365,169 @@ static int
 DropShards(Oid relationId, char *schemaName, char *relationName,
 		   List *deletableShardIntervalList)
 {
-	ListCell *shardIntervalCell = NULL;
+	Assert(OidIsValid(relationId));
+	Assert(schemaName != NULL);
+	Assert(relationName != NULL);
 
 	UseCoordinatedTransaction();
 
-	/* At this point we intentionally decided to not use 2PC for reference tables */
+	/*
+	 * At this point we intentionally decided to not use 2PC for reference
+	 * tables
+	 */
 	if (MultiShardCommitProtocol == COMMIT_PROTOCOL_2PC)
 	{
 		CoordinatedTransactionUse2PC();
 	}
 
-	foreach(shardIntervalCell, deletableShardIntervalList)
+	ShardInterval *shardInterval = NULL;
+
+	foreach_ptr(shardInterval, deletableShardIntervalList)
 	{
-		ListCell *shardPlacementCell = NULL;
-		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
 		uint64 shardId = shardInterval->shardId;
 		char *shardRelationName = pstrdup(relationName);
 
 		Assert(shardInterval->relationId == relationId);
 
-		/* Build shard relation name. */
+		/* build shard relation name */
 		AppendShardIdToName(&shardRelationName, shardId);
-		char *quotedShardName = quote_qualified_identifier(schemaName, shardRelationName);
 
 		List *shardPlacementList = ShardPlacementList(shardId);
-		foreach(shardPlacementCell, shardPlacementList)
+
+		ShardPlacement *shardPlacement = NULL;
+
+		foreach_ptr(shardPlacement, shardPlacementList)
 		{
-			ShardPlacement *shardPlacement =
-				(ShardPlacement *) lfirst(shardPlacementCell);
-			char *workerName = shardPlacement->nodeName;
-			uint32 workerPort = shardPlacement->nodePort;
-			StringInfo workerDropQuery = makeStringInfo();
-			uint32 connectionFlags = FOR_DDL;
-
-			char storageType = shardInterval->storageType;
-			if (storageType == SHARD_STORAGE_TABLE)
+			uint64 shardPlacementId = shardPlacement->placementId;
+
+			if (shardPlacement->groupId == COORDINATOR_GROUP_ID &&
+				IsCoordinator() &&
+				DropSchemaOrDBInProgress())
 			{
-				appendStringInfo(workerDropQuery, DROP_REGULAR_TABLE_COMMAND,
-								 quotedShardName);
+				/*
+				 * The active DROP SCHEMA/DATABASE ... CASCADE will drop the
+				 * shard, if we try to drop it over another connection, we will
+				 * get into a distributed deadlock.
+				 */
 			}
-			else if (storageType == SHARD_STORAGE_COLUMNAR ||
-					 storageType == SHARD_STORAGE_FOREIGN)
+			else
 			{
-				appendStringInfo(workerDropQuery, DROP_FOREIGN_TABLE_COMMAND,
-								 quotedShardName);
+				char storageType = shardInterval->storageType;
+
+				/* build command to drop the shard */
+				const char *dropShardPlacementCommand = CreateDropShardPlacementCommand(
+					schemaName, shardRelationName, storageType);
+
+				/*
```

Comment:
Minor: I think the function name pretty clear, so the comment seems a little unnecessary, consider removing it